### DISCUSSION TITLE
Could not load file or assembly CefSharp

### DISCUSSION THREAD

**Discussion by phuocminh6464** [cefsharp/CefSharp#4352] [General]
I  have a WPF net core project and it's using CEF ChromiumWebBrowser, in the debug mode it run ok. Then I create Setup Project to make file msi file for client but it has error "File 'CefSharp.Core.Runtime.dll' of project output 'Publish Items from CommunicateWithWeb (Active)' targeting 'x64' is not compatible with the project's target platform 'x86'" while building. After I change TargetPlatform of Setup Project to x64, build is successfully. Then I run msi file to install app in my PC but I can't open this app by clicking exe file. I open event viewer and error is like picture below.

![image](https://user-images.githubusercontent.com/26184798/209895809-76c0ba6b-cf34-4e76-8753-ebaad8f43a77.png)

I tried to installed Orca and see that CefSharp assembly is already included in .msi file as picture below

![image](https://user-images.githubusercontent.com/26184798/209895874-b02dcce0-1152-421d-b9b1-adaddcc0ef42.png)

Anyone can help me this problem? Any comment is also important for me. Thank you.

Note: I use VS2022 64bit, Win10 64bit, Net 6.0, CefSharp.WPF.NETCore ver 108.4.130. I also install MS Visual C++ 2015-2022 Redistributable (both x64 and x86) ver 14.34.31931. I set platform = "x64" when build Setup Project in Release mode

**Comment by amaitland** [comment]
In future please use [Discussions](https://github.com/cefsharp/CefSharp/discussions) to ask support related questions. 

Looks like you've already asked this at https://stackoverflow.com/questions/74927986/could-not-load-file-or-assembly-cefsharp
I'd also ask that you ask your question just **once** and wait for a reply. Posting in multiple locations creates extra work. You need to post a link on your `stackoverflow` question to this discussion now.

> I tried to installed Orca and see that CefSharp assembly is already included in .msi file as picture below

Did you check your installer contains all the required files? https://github.com/cefsharp/CefSharp/blob/master/NuGet/PackageReference/Readme.txt#L25

There's also a list at https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29

Have you checked the installation folder to confirm that the files are actually there? This will likely be easier than digging through the msi.

> Net 6.0, CefSharp.WPF.NETCore ver 108.4.130

Did you set a [RuntimeIdentifier](https://github.com/cefsharp/CefSharp/wiki/Quick-Start-For-MS-.Net-5.0-or-greater#20-set-a-runtimeidentifier-in-your-proj-file) as per the [post install instructions](https://github.com/cefsharp/CefSharp/blob/master/NuGet/PackageReference/Readme.txt#L10-L11)?

If you've installed `VC++` then the most likely problem is your installer is missing some of the dependencies.

> Then I create Setup Project to make file msi file for client but it has error "File 'CefSharp.Core.Runtime.dll' of project output 'Publish Items from CommunicateWithWeb (Active)'

If you've set a `RuntimeIdentifier` and still having problems then you'll need to provide an example that reproduces the problem. You can fork the https://github.com/cefsharp/CefSharp.MinimalExample and push your changes to `GitHub` (no zip files, I won't open them).

**Reply by phuocminh6464** [reply]
Hi @amaitland thank for your help. I can open my app after I set a RuntimeIdentifier in my proj file as the link you give me, but I have another problem about load chromiumwebbrowser.
What I do: Open wpf app then click the button which will display a web page, but app will close immediately when I click that button. I open file debug.log and see the below error
![image](https://user-images.githubusercontent.com/26184798/210296020-aafcd536-c4f1-4399-8394-2c32784b4e29.png)

I open viewer and see the error
![image](https://user-images.githubusercontent.com/26184798/210296124-fcfdaedc-c7da-44ca-81dd-fa3cd66c71a4.png)
![image](https://user-images.githubusercontent.com/26184798/210296310-01f39cb2-fb47-4ae7-81f5-d87f959331b6.png)

I also checked and my installation folder contained all required dependencies as link https://github.com/cefsharp/CefSharp/blob/master/NuGet/PackageReference/Readme.txt#L25

So I tried to add try catch block for click event and also wrote log before each code line, but program run entirely code line without error so catch block is not called (click event only open web page by using ChromiumWebBrowser.Load() method). Could you help me with this problem? My code run normally in debug and release mode in visual studio, but it only failed with .msi file.

**Reply by amaitland** [reply]
> What I do: Open wpf app then click the button which will display a web page, but app will close immediately when I click that button. I open file debug.log and see the below error

Post your errors as formatted text please.

Also include the button code, just saying you click a button and it crashes isn't enough information.

> So I tried to add try catch block for click event and also wrote log before each code line, but program run entirely code line without error so catch block is not called (click event only open web page by using ChromiumWebBrowser.Load() method)

I have no idea what you mean by this.

**Reply by phuocminh6464** [reply]
This is content of debug.log file from I open app to it crash after click button:

```
[0103/131641.062:VERBOSE1:pref_proxy_config_tracker_impl.cc(187)] 406D0700: set chrome proxy config service to 40660800
[0103/131641.073:VERBOSE1:webrtc_internals.cc(120)] Could not get the download directory.
[0103/131641.073:VERBOSE1:media_stream_manager.cc(1100)] MSM::InitializeMaybeAsync([this=40691500])
[0103/131641.074:VERBOSE1:media_stream_manager.cc(1100)] MDM::MediaDevicesManager()
[0103/131641.074:VERBOSE1:media_stream_manager.cc(1100)] MSM::MediaStreamManager([this=40691500]))
[0103/131641.087:VERBOSE1:component_installer.cc(285)] StartRegistration for Widevine Content Decryption Module
[0103/131641.087:VERBOSE1:first_party_sets_handler_impl.cc(279)] Empty path. Failed initializing First-Party Sets database.
[0103/131641.088:VERBOSE1:component_installer.cc(439)] FinishRegistration for Widevine Content Decryption Module
[0103/131641.088:VERBOSE1:component_updater_service.cc(131)] CrxUpdateService starting up. First update attempt will take place in 60 seconds. Next update attempt will take place in 18000 seconds. 
[0103/131641.530:VERBOSE1:configured_proxy_resolution_service.cc(1095)] Failed configuring with PAC script, falling-back to manual proxy servers.
[0103/131643.128:VERBOSE1:pref_proxy_config_tracker_impl.cc(187)] 406D1180: set chrome proxy config service to 4079EF00
[0103/131643.128:WARNING:account_consistency_mode_manager.cc(70)] Desktop Identity Consistency cannot be enabled as no OAuth client ID and client secret have been configured.
[0103/131643.128:VERBOSE1:account_consistency_mode_manager.cc(197)] Desktop Identity Consistency disabled as sign-in to Chrome is not allowed
[0103/131643.129:VERBOSE1:account_consistency_mode_manager.cc(197)] Desktop Identity Consistency disabled as sign-in to Chrome is not allowed
[0103/131643.129:VERBOSE1:mutable_profile_oauth2_token_service_delegate.cc(207)] MutablePO2TS::MutablePO2TS
[0103/131643.143:VERBOSE1:video_capture_oracle.cc(91)] Capture size auto-throttling is now enabled.
[0103/131643.144:VERBOSE1:video_capture_oracle.cc(120)] Capture size auto-throttling is now disabled.
[0103/131643.144:VERBOSE1:capture_resolution_chooser.cc(205)] Recomputed snapped frame sizes: 600x411
[0103/131643.144:VERBOSE1:capture_resolution_chooser.cc(157)] Recomputed capture size from 640x360 to 600x411 (100% of ideal size)
[0103/131643.152:WARNING:alloy_constrained_window_views_client.cc(63)] No likely focused browser
[0103/131643.156:VERBOSE1:configured_proxy_resolution_service.cc(1095)] Failed configuring with PAC script, falling-back to manual proxy servers.
[0103/131643.340:INFO:cpu_info.cc(53)] Available number of cores: 4
[0103/131643.357:INFO:cpu_info.cc(53)] Available number of cores: 4
```

And this is code for button:
```c#
private void Button_Click(object sender, RoutedEventArgs e)
{
	try
	{
		LogHelper.LogInfo("Before Load");
		web.JavascriptObjectRepository.UnRegisterAll();
		web.JavascriptObjectRepository.Register("wpfObj", new JavaScriptInteractionObj(), options: BindingOptions.DefaultBinder);
		web.Load(urlLoad);  
		LogHelper.LogInfo("After Load");
		
		LogHelper.LogInfo("Before JavascriptMessageReceived");
		web.JavascriptMessageReceived -= OnBrowserJavascriptMessageReceived; 
		web.JavascriptMessageReceived += OnBrowserJavascriptMessageReceived; 
		LogHelper.LogInfo("After JavascriptMessageReceived");
	}
	catch (Exception ex)
	{
		LogHelper.LogError(ex.Message);
	}
}
```

LogHelper is a class that I create with using log4net to see the detail error. When app crash, I open my log file and the last line is "After JavascriptMessageReceived" . It mean this app run success without error so I really don't know why app crash

**Reply by amaitland** [reply]
> [0103/131643.152:WARNING:alloy_constrained_window_views_client.cc(63)] No likely focused browser

This is a likely candidate for the cause of the crash. Try calling `web.Focus()` before your `web.Load` call.

There's a discussion at https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=19318

You won't be able to catch the exception as it comes from one of the `CEF` unmanaged threads.

> but it only failed with .msi file.

Does the crash happen on your `Dev` machine? Or a different machine?

**Reply by phuocminh6464** [reply]
> This is a likely candidate for the cause of the crash. Try calling web.Focus() before your web.Load call.

I called Focus method but app still crash.

> There's a discussion at https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=19318
You won't be able to catch the exception as it comes from one of the CEF unmanaged threads.

I read that discussion and trying share RequestContext settings with GlobalRequestContext but it's not working. However, when I change CEF ver 108 (I used 108.4.13) to CEF ver 107.1.4 my app is working perfectly. So I think this is problem of CEF ver 108. Could you help me check it?

> Does the crash happen on your Dev machine? Or a different machine?

This app is working right in my PC (dev environment both debug and release mode) and other dev machines, it only crash when I build to .msi file by using Setup project in Visual Studio and then open .exe file after installing .msi file

**Reply by amaitland** [reply]
> So I think this is problem of CEF ver 108

That is correct. The bug will need to be fixed in `CEF`, not something I can fix in `CefSharp` unfortunately.

> it only crash when I build to .msi file by using Setup project in Visual Studio and then open .exe file after installing .msi file

Does the `msi` install your application into `Program Files`? If yes try copy the installed application into a temp directory that has write permissions.

**Reply by phuocminh6464** [reply]
@amaitland Thank you so much. I can solve all my problem after changing to use CEF ver 107 with the original source

**Reply by amaitland** [reply]
> Does the msi install your application into Program Files? If yes try copy the installed application into a temp directory that has write permissions.

Can you answer my question and perform the test if appropriate please.

Please use version `108` for the testing. The sooner I understand how to reproduce the problem the better.

**Reply by phuocminh6464** [reply]
> Does the msi install your application into Program Files? If yes try copy the installed application into a temp directory that has write permissions.

Yes, the msi install my application into `Program Files` folder.  You are right, when I copy the installed application into a temp directory that has write permissions (Disk D), app is working normally (I'm testing with version 108.4.13)

**Reply by amaitland** [reply]
> Yes, the msi install my application into `Program Files` folder. You are right, when I copy the installed application into a temp directory that has write permissions (Disk D), app is working normally (I'm testing with version 108.4.13)

Thank you. Very helpful.

Did you set `CefSettings.CachePath`? I suspect the problem only occurs when no `CachePath` is set and the application folder is read only.

There's an example at https://github.com/cefsharp/CefSharp/wiki/General-Usage#initialize-and-shutdown

**Reply by phuocminh6464** [reply]
Perfectly, this is my fault. After I set CachePath in `App.xaml.cs` file, app is working right now with ver 108 and install files are still in `Program Files` folder

**Reply by amaitland** [reply]
> Perfectly, this is my fault

It's a bug in `CEF`. It should work without setting a `CachePath`.

I can now reproduce the issue reliably using the `CEF Sample Application(cefclient)` and have reporting my findings at https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=19318#p53037

Thanks for testing 👍
